// objection/lib/relations/belongsToOne/BelongsToOneRelateOperation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _normalizeIds = require('../../utils/normalizeIds');

var _normalizeIds2 = _interopRequireDefault(_normalizeIds);

var _QueryBuilderOperation = require('../../queryBuilder/operations/QueryBuilderOperation');

var _QueryBuilderOperation2 = _interopRequireDefault(_QueryBuilderOperation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BelongsToOneRelateOperation = function (_QueryBuilderOperatio) {
  (0, _inherits3.default)(BelongsToOneRelateOperation, _QueryBuilderOperatio);

  function BelongsToOneRelateOperation(builder, name, opt) {
    (0, _classCallCheck3.default)(this, BelongsToOneRelateOperation);

    var _this = (0, _possibleConstructorReturn3.default)(this, _QueryBuilderOperatio.call(this, builder, name, opt));

    _this.isWriteOperation = true;
    _this.relation = opt.relation;
    _this.owner = opt.owner;
    _this.input = null;
    _this.ids = null;
    return _this;
  }

  BelongsToOneRelateOperation.prototype.call = function call(builder, args) {
    this.input = args[0];
    this.ids = (0, _normalizeIds2.default)(args[0], this.relation.relatedProp, { arrayOutput: true });

    if (this.ids.length > 1) {
      this.relation.throwError('can only relate one model to a BelongsToOneRelation');
    }

    return true;
  };

  BelongsToOneRelateOperation.prototype.queryExecutor = function queryExecutor(builder) {
    var _this2 = this;

    var patch = {};

    _lodash2.default.each(this.relation.ownerProp, function (prop, idx) {
      patch[prop] = _this2.ids[0][idx];
      _this2.owner[prop] = _this2.ids[0][idx];
    });

    return this.relation.ownerModelClass.query().childQueryOf(builder).copyFrom(builder, /where/i).patch(patch).whereComposite(this.relation.ownerModelClass.getFullIdColumn(), this.owner.$id());
  };

  BelongsToOneRelateOperation.prototype.onAfterInternal = function onAfterInternal(builder) {
    return this.input;
  };

  return BelongsToOneRelateOperation;
}(_QueryBuilderOperation2.default);

exports.default = BelongsToOneRelateOperation;
