// objection/lib/relations/belongsToOne/BelongsToOneRelation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Relation2 = require('../Relation');

var _Relation3 = _interopRequireDefault(_Relation2);

var _BelongsToOneInsertOperation = require('./BelongsToOneInsertOperation');

var _BelongsToOneInsertOperation2 = _interopRequireDefault(_BelongsToOneInsertOperation);

var _BelongsToOneRelateOperation = require('./BelongsToOneRelateOperation');

var _BelongsToOneRelateOperation2 = _interopRequireDefault(_BelongsToOneRelateOperation);

var _BelongsToOneUnrelateOperation = require('./BelongsToOneUnrelateOperation');

var _BelongsToOneUnrelateOperation2 = _interopRequireDefault(_BelongsToOneUnrelateOperation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BelongsToOneRelation = function (_Relation) {
  (0, _inherits3.default)(BelongsToOneRelation, _Relation);

  function BelongsToOneRelation() {
    (0, _classCallCheck3.default)(this, BelongsToOneRelation);
    return (0, _possibleConstructorReturn3.default)(this, _Relation.apply(this, arguments));
  }

  BelongsToOneRelation.prototype.createRelationProp = function createRelationProp(owners, related) {
    var _this2 = this;

    var relatedByOwnerId = _lodash2.default.keyBy(related, function (related) {
      return related.$values(_this2.relatedProp);
    });

    _lodash2.default.each(owners, function (owner) {
      var ownerId = owner.$values(_this2.ownerProp);
      owner[_this2.name] = relatedByOwnerId[ownerId] || null;
    });
  };

  BelongsToOneRelation.prototype.insert = function insert(builder, owner) {
    return new _BelongsToOneInsertOperation2.default(builder, 'insert', {
      relation: this,
      owner: owner
    });
  };

  BelongsToOneRelation.prototype.relate = function relate(builder, owner) {
    return new _BelongsToOneRelateOperation2.default(builder, 'relate', {
      relation: this,
      owner: owner
    });
  };

  BelongsToOneRelation.prototype.unrelate = function unrelate(builder, owner) {
    return new _BelongsToOneUnrelateOperation2.default(builder, 'unrelate', {
      relation: this,
      owner: owner
    });
  };

  return BelongsToOneRelation;
}(_Relation3.default);

exports.default = BelongsToOneRelation;
