// objection/lib/relations/manyToMany/ManyToManyRelateOperation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _normalizeIds = require('../../utils/normalizeIds');

var _normalizeIds2 = _interopRequireDefault(_normalizeIds);

var _QueryBuilderOperation = require('../../queryBuilder/operations/QueryBuilderOperation');

var _QueryBuilderOperation2 = _interopRequireDefault(_QueryBuilderOperation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ManyToManyRelateOperation = function (_QueryBuilderOperatio) {
  (0, _inherits3.default)(ManyToManyRelateOperation, _QueryBuilderOperatio);

  function ManyToManyRelateOperation(builder, name, opt) {
    (0, _classCallCheck3.default)(this, ManyToManyRelateOperation);

    var _this = (0, _possibleConstructorReturn3.default)(this, _QueryBuilderOperatio.call(this, builder, name, opt));

    _this.isWriteOperation = true;
    _this.relation = opt.relation;
    _this.owner = opt.owner;
    _this.input = null;
    _this.ids = null;
    return _this;
  }

  ManyToManyRelateOperation.prototype.call = function call(builder, args) {
    this.input = args[0];
    this.ids = (0, _normalizeIds2.default)(args[0], this.relation.relatedProp);
    return true;
  };

  ManyToManyRelateOperation.prototype.queryExecutor = function queryExecutor(builder) {
    var joinModels = this.relation.createJoinModels(this.owner.$values(this.relation.ownerProp), this.ids);

    return this.relation.joinTableModelClass.bindKnex(builder.knex()).query().childQueryOf(builder).insert(joinModels);
  };

  ManyToManyRelateOperation.prototype.onAfterInternal = function onAfterInternal() {
    return this.input;
  };

  return ManyToManyRelateOperation;
}(_QueryBuilderOperation2.default);

exports.default = ManyToManyRelateOperation;
