// objection/lib/relations/manyToMany/ManyToManyRelation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _desc, _value, _class;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Relation2 = require('../Relation');

var _Relation3 = _interopRequireDefault(_Relation2);

var _ModelBase = require('../../model/ModelBase');

var _ModelBase2 = _interopRequireDefault(_ModelBase);

var _inheritModel = require('../../model/inheritModel');

var _inheritModel2 = _interopRequireDefault(_inheritModel);

var _normalizeIds = require('../../utils/normalizeIds');

var _normalizeIds2 = _interopRequireDefault(_normalizeIds);

var _dbUtils = require('../../utils/dbUtils');

var _classUtils = require('../../utils/classUtils');

var _memoize = require('../../utils/decorators/memoize');

var _memoize2 = _interopRequireDefault(_memoize);

var _ManyToManyFindOperation = require('./ManyToManyFindOperation');

var _ManyToManyFindOperation2 = _interopRequireDefault(_ManyToManyFindOperation);

var _ManyToManyInsertOperation = require('./ManyToManyInsertOperation');

var _ManyToManyInsertOperation2 = _interopRequireDefault(_ManyToManyInsertOperation);

var _ManyToManyRelateOperation = require('./ManyToManyRelateOperation');

var _ManyToManyRelateOperation2 = _interopRequireDefault(_ManyToManyRelateOperation);

var _ManyToManyUnrelateOperation = require('./ManyToManyUnrelateOperation');

var _ManyToManyUnrelateOperation2 = _interopRequireDefault(_ManyToManyUnrelateOperation);

var _ManyToManyUnrelateSqliteOperation = require('./ManyToManyUnrelateSqliteOperation');

var _ManyToManyUnrelateSqliteOperation2 = _interopRequireDefault(_ManyToManyUnrelateSqliteOperation);

var _ManyToManyUpdateOperation = require('./ManyToManyUpdateOperation');

var _ManyToManyUpdateOperation2 = _interopRequireDefault(_ManyToManyUpdateOperation);

var _ManyToManyUpdateSqliteOperation = require('./ManyToManyUpdateSqliteOperation');

var _ManyToManyUpdateSqliteOperation2 = _interopRequireDefault(_ManyToManyUpdateSqliteOperation);

var _ManyToManyDeleteOperation = require('./ManyToManyDeleteOperation');

var _ManyToManyDeleteOperation2 = _interopRequireDefault(_ManyToManyDeleteOperation);

var _ManyToManyDeleteSqliteOperation = require('./ManyToManyDeleteSqliteOperation');

var _ManyToManyDeleteSqliteOperation2 = _interopRequireDefault(_ManyToManyDeleteSqliteOperation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var sqliteBuiltInRowId = '_rowid_';

var ManyToManyRelation = (_class = function (_Relation) {
  (0, _inherits3.default)(ManyToManyRelation, _Relation);

  function ManyToManyRelation() {
    (0, _classCallCheck3.default)(this, ManyToManyRelation);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    /**
     * @type {string}
     */

    var _this = (0, _possibleConstructorReturn3.default)(this, _Relation.call.apply(_Relation, [this].concat(args)));

    _this.joinTable = null;

    /**
     * @type {Array.<string>}
     */
    _this.joinTableOwnerCol = null;

    /**
     * @type {Array.<string>}
     */
    _this.joinTableOwnerProp = null;

    /**
     * @type {Array.<string>}
     */
    _this.joinTableRelatedCol = null;

    /**
     * @type {Array.<string>}
     */
    _this.joinTableRelatedProp = null;

    /**
     * @type {Constructor.<Model>}
     */
    _this.joinTableModelClass = null;

    /**
     * @type {Array.<string>}
     */
    _this.joinTableExtraCols = null;

    /**
     * @type {Array.<string>}
     */
    _this.joinTableExtraProps = null;
    return _this;
  }

  ManyToManyRelation.prototype.setMapping = function setMapping(mapping) {
    var retVal = _Relation.prototype.setMapping.call(this, mapping);

    // Avoid require loop and import here.
    var Model = require(__dirname + '/../../model/Model').default;

    if (!_lodash2.default.isObject(mapping.join.through)) {
      this.throwError('join must have the `through` that describes the join table.');
    }

    if (!mapping.join.through.from || !mapping.join.through.to) {
      this.throwError('join.through must be an object that describes the join table. For example: {from: "JoinTable.someId", to: "JoinTable.someOtherId"}');
    }

    var joinFrom = this.parseReference(mapping.join.from);
    var joinTableFrom = this.parseReference(mapping.join.through.from);
    var joinTableTo = this.parseReference(mapping.join.through.to);
    var joinTableExtra = mapping.join.through.extra || [];

    if (!joinTableFrom.table || _lodash2.default.isEmpty(joinTableFrom.columns)) {
      this.throwError('join.through.from must have format JoinTable.columnName. For example "JoinTable.someId" or in case of composite key ["JoinTable.a", "JoinTable.b"].');
    }

    if (!joinTableTo.table || _lodash2.default.isEmpty(joinTableTo.columns)) {
      this.throwError('join.through.to must have format JoinTable.columnName. For example "JoinTable.someId" or in case of composite key ["JoinTable.a", "JoinTable.b"].');
    }

    if (joinTableFrom.table !== joinTableTo.table) {
      this.throwError('join.through `from` and `to` must point to the same join table.');
    }

    this.joinTable = joinTableFrom.table;
    this.joinTableExtraCols = joinTableExtra;

    if (joinFrom.table === this.ownerModelClass.tableName) {
      this.joinTableOwnerCol = joinTableFrom.columns;
      this.joinTableRelatedCol = joinTableTo.columns;
    } else {
      this.joinTableRelatedCol = joinTableFrom.columns;
      this.joinTableOwnerCol = joinTableTo.columns;
    }

    if (mapping.join.through.modelClass) {
      var modelClass = mapping.join.through.modelClass;

      try {
        if (_lodash2.default.isString(modelClass)) {
          modelClass = require(modelClass);

          // Compatibility with babel `export default`.
          if (!(0, _classUtils.isSubclassOf)(modelClass, Model) && (0, _classUtils.isSubclassOf)(modelClass.default, Model)) {
            modelClass = modelClass.default;
          }
        }
      } catch (err) {
        // Do nothing.
      }

      if (!(0, _classUtils.isSubclassOf)(modelClass, Model)) {
        this.throwError('Join table model class is not a subclass of Model');
      }

      this.joinTableModelClass = modelClass;
    } else {
      this.joinTableModelClass = (0, _inheritModel2.default)(Model);
      this.joinTableModelClass.tableName = this.joinTable;
      // We cannot know if the join table has a primary key. Therefore we set some
      // known column as the idColumn so that inserts will work.
      this.joinTableModelClass.idColumn = this.joinTableRelatedCol;
    }

    this.joinTableOwnerProp = this.propertyName(this.joinTableOwnerCol, this.joinTableModelClass);
    this.joinTableRelatedProp = this.propertyName(this.joinTableRelatedCol, this.joinTableModelClass);
    this.joinTableExtraProps = this.propertyName(this.joinTableExtraCols, this.joinTableModelClass);

    return retVal;
  };

  /**
   * @returns {Array.<string>}
   */


  ManyToManyRelation.prototype.fullJoinTableOwnerCol = function fullJoinTableOwnerCol() {
    var _this2 = this;

    return _lodash2.default.map(this.joinTableOwnerCol, function (col) {
      return _this2.joinTable + '.' + col;
    });
  };

  /**
   * @returns {Array.<string>}
   */


  ManyToManyRelation.prototype.fullJoinTableRelatedCol = function fullJoinTableRelatedCol() {
    var _this3 = this;

    return _lodash2.default.map(this.joinTableRelatedCol, function (col) {
      return _this3.joinTable + '.' + col;
    });
  };

  /**
   * @returns {Array.<string>}
   */


  ManyToManyRelation.prototype.fullJoinTableExtraCols = function fullJoinTableExtraCols() {
    var _this4 = this;

    return _lodash2.default.map(this.joinTableExtraCols, function (col) {
      return _this4.joinTable + '.' + col;
    });
  };

  /**
   * @returns {string}
   */


  ManyToManyRelation.prototype.joinTableAlias = function joinTableAlias() {
    return this.joinTable + '_rel_' + this.name;
  };

  /**
   * @returns {ManyToManyRelation}
   */


  ManyToManyRelation.prototype.clone = function clone() {
    var relation = _Relation.prototype.clone.call(this);

    relation.joinTable = this.joinTable;
    relation.joinTableOwnerCol = this.joinTableOwnerCol;
    relation.joinTableOwnerProp = this.joinTableOwnerProp;
    relation.joinTableRelatedCol = this.joinTableRelatedCol;
    relation.joinTableRelatedProp = this.joinTableRelatedProp;
    relation.joinTableModelClass = this.joinTableModelClass;
    relation.joinTableExtraCols = this.joinTableExtraCols;
    relation.joinTableExtraProps = this.joinTableExtraProps;

    return relation;
  };

  /**
   * @returns {ManyToManyRelation}
   */


  ManyToManyRelation.prototype.bindKnex = function bindKnex(knex) {
    var bound = _Relation.prototype.bindKnex.call(this, knex);
    bound.joinTableModelClass = this.joinTableModelClass.bindKnex(knex);
    return bound;
  };

  /**
   * @returns {QueryBuilder}
   */


  ManyToManyRelation.prototype.findQuery = function findQuery(builder, ownerIds, isColumnRef) {
    var _this5 = this;

    var fullRelatedCol = this.fullRelatedCol();

    builder.join(this.joinTable, function (join) {
      _lodash2.default.each(_this5.fullJoinTableRelatedCol(), function (joinTableRelatedCol, idx) {
        join.on(joinTableRelatedCol, fullRelatedCol[idx]);
      });
    });

    if (isColumnRef) {
      _lodash2.default.each(this.fullJoinTableOwnerCol(), function (joinTableOwnerCol, idx) {
        builder.whereRef(joinTableOwnerCol, ownerIds[idx]);
      });
    } else {
      if ((0, _lodash2.default)(ownerIds).flatten().every(function (id) {
        return _lodash2.default.isNull(id) || _lodash2.default.isUndefined(id);
      })) {
        // Nothing to fetch.
        builder.resolve([]);
      } else {
        builder.whereInComposite(this.fullJoinTableOwnerCol(), ownerIds);
      }
    }

    return builder.modify(this.filter);
  };

  /**
   * @returns {QueryBuilder}
   */


  ManyToManyRelation.prototype.join = function join(builder, joinOperation, relatedTableAlias) {
    joinOperation = joinOperation || 'join';
    relatedTableAlias = relatedTableAlias || this.relatedTableAlias();

    var joinTable = this.joinTable;
    var relatedTable = this.relatedModelClass.tableName;

    var joinTableAlias = this.joinTableAlias();
    var joinTableAsAlias = joinTable + ' as ' + joinTableAlias;
    var relatedTableAsAlias = relatedTable + ' as ' + relatedTableAlias;

    var joinTableOwnerCol = _lodash2.default.map(this.joinTableOwnerCol, function (col) {
      return joinTableAlias + '.' + col;
    });
    var joinTableRelatedCol = _lodash2.default.map(this.joinTableRelatedCol, function (col) {
      return joinTableAlias + '.' + col;
    });

    var ownerCol = this.fullOwnerCol();
    var relatedCol = _lodash2.default.map(this.relatedCol, function (col) {
      return relatedTableAlias + '.' + col;
    });

    return builder[joinOperation](joinTableAsAlias, function (join) {
      _lodash2.default.each(joinTableOwnerCol, function (joinTableOwnerCol, idx) {
        join.on(joinTableOwnerCol, ownerCol[idx]);
      });
    })[joinOperation](relatedTableAsAlias, function (join) {
      _lodash2.default.each(joinTableRelatedCol, function (joinTableRelatedCol, idx) {
        join.on(joinTableRelatedCol, relatedCol[idx]);
      });
    }).modify(this.filter);
  };

  ManyToManyRelation.prototype.find = function find(builder, owners) {
    return new _ManyToManyFindOperation2.default(builder, 'find', {
      relation: this,
      owners: owners
    });
  };

  ManyToManyRelation.prototype.insert = function insert(builder, owner) {
    return new _ManyToManyInsertOperation2.default(builder, 'insert', {
      relation: this,
      owner: owner
    });
  };

  ManyToManyRelation.prototype.update = function update(builder, owner) {
    if ((0, _dbUtils.isSqlite)(builder.knex())) {
      return new _ManyToManyUpdateSqliteOperation2.default(builder, 'update', {
        relation: this,
        owner: owner
      });
    } else {
      return new _ManyToManyUpdateOperation2.default(builder, 'update', {
        relation: this,
        owner: owner
      });
    }
  };

  ManyToManyRelation.prototype.patch = function patch(builder, owner) {
    if ((0, _dbUtils.isSqlite)(builder.knex())) {
      return new _ManyToManyUpdateSqliteOperation2.default(builder, 'patch', {
        relation: this,
        owner: owner,
        modelOptions: { patch: true }
      });
    } else {
      return new _ManyToManyUpdateOperation2.default(builder, 'patch', {
        relation: this,
        owner: owner,
        modelOptions: { patch: true }
      });
    }
  };

  ManyToManyRelation.prototype.delete = function _delete(builder, owner) {
    if ((0, _dbUtils.isSqlite)(builder.knex())) {
      return new _ManyToManyDeleteSqliteOperation2.default(builder, 'delete', {
        relation: this,
        owner: owner
      });
    } else {
      return new _ManyToManyDeleteOperation2.default(builder, 'delete', {
        relation: this,
        owner: owner
      });
    }
  };

  ManyToManyRelation.prototype.relate = function relate(builder, owner) {
    return new _ManyToManyRelateOperation2.default(builder, 'relate', {
      relation: this,
      owner: owner
    });
  };

  ManyToManyRelation.prototype.unrelate = function unrelate(builder, owner) {
    if ((0, _dbUtils.isSqlite)(builder.knex())) {
      return new _ManyToManyUnrelateSqliteOperation2.default(builder, 'unrelate', {
        relation: this,
        owner: owner
      });
    } else {
      return new _ManyToManyUnrelateOperation2.default(builder, 'unrelate', {
        relation: this,
        owner: owner
      });
    }
  };

  ManyToManyRelation.prototype.selectForModify = function selectForModify(builder, owner) {
    var ownerId = owner.$values(this.ownerProp);

    var idQuery = this.joinTableModelClass.bindKnex(builder.knex()).query().childQueryOf(builder).select(this.fullJoinTableRelatedCol()).whereComposite(this.fullJoinTableOwnerCol(), ownerId);

    return builder.whereInComposite(this.fullRelatedCol(), idQuery);
  };

  ManyToManyRelation.prototype.selectForModifySqlite = function selectForModifySqlite(builder, owner) {
    var _this6 = this;

    var relatedTable = this.relatedModelClass.tableName;
    var relatedTableAlias = this.relatedTableAlias();
    var relatedTableAsAlias = relatedTable + ' as ' + relatedTableAlias;
    var relatedTableAliasRowId = relatedTableAlias + '.' + sqliteBuiltInRowId;
    var relatedTableRowId = relatedTable + '.' + sqliteBuiltInRowId;

    var fullRelatedCol = this.fullRelatedCol();
    var ownerId = owner.$values(this.ownerProp);

    var selectRelatedQuery = this.joinTableModelClass.bindKnex(builder.knex()).query().childQueryOf(builder).select(relatedTableAliasRowId).whereComposite(this.fullJoinTableOwnerCol(), ownerId).join(relatedTableAsAlias, function (join) {
      _lodash2.default.each(_this6.fullJoinTableRelatedCol(), function (joinTableRelatedCol, idx) {
        join.on(joinTableRelatedCol, fullRelatedCol[idx]);
      });
    });

    return builder.whereInComposite(relatedTableRowId, selectRelatedQuery);
  };

  ManyToManyRelation.prototype.createJoinModels = function createJoinModels(ownerId, related) {
    var _this7 = this;

    return _lodash2.default.map(related, function (related) {
      var joinModel = {};

      _lodash2.default.each(_this7.joinTableOwnerProp, function (joinTableOwnerProp, idx) {
        joinModel[joinTableOwnerProp] = ownerId[idx];
      });

      _lodash2.default.each(_this7.joinTableRelatedProp, function (joinTableRelatedProp, idx) {
        joinModel[joinTableRelatedProp] = related[_this7.relatedProp[idx]];
      });

      _lodash2.default.each(_this7.joinTableExtraProps, function (extraProp) {
        if (!_lodash2.default.isUndefined(related[extraProp])) {
          joinModel[extraProp] = related[extraProp];
        }
      });

      return joinModel;
    });
  };

  ManyToManyRelation.prototype.omitExtraProps = function omitExtraProps(models) {
    var _this8 = this;

    if (!_lodash2.default.isEmpty(this.joinTableExtraProps)) {
      _lodash2.default.each(models, function (model) {
        return model.$omitFromDatabaseJson(_this8.joinTableExtraProps);
      });
    }
  };

  return ManyToManyRelation;
}(_Relation3.default), (_applyDecoratedDescriptor(_class.prototype, 'fullJoinTableOwnerCol', [_memoize2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'fullJoinTableOwnerCol'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'fullJoinTableRelatedCol', [_memoize2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'fullJoinTableRelatedCol'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'fullJoinTableExtraCols', [_memoize2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'fullJoinTableExtraCols'), _class.prototype)), _class);
exports.default = ManyToManyRelation;
