// objection/lib/relations/hasMany/HasManyRelateOperation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _normalizeIds = require('../../utils/normalizeIds');

var _normalizeIds2 = _interopRequireDefault(_normalizeIds);

var _QueryBuilderOperation = require('../../queryBuilder/operations/QueryBuilderOperation');

var _QueryBuilderOperation2 = _interopRequireDefault(_QueryBuilderOperation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var HasManyRelateOperation = function (_QueryBuilderOperatio) {
  (0, _inherits3.default)(HasManyRelateOperation, _QueryBuilderOperatio);

  function HasManyRelateOperation(builder, name, opt) {
    (0, _classCallCheck3.default)(this, HasManyRelateOperation);

    var _this = (0, _possibleConstructorReturn3.default)(this, _QueryBuilderOperatio.call(this, builder, name, opt));

    _this.isWriteOperation = true;
    _this.relation = opt.relation;
    _this.owner = opt.owner;
    _this.input = null;
    _this.ids = null;
    return _this;
  }

  HasManyRelateOperation.prototype.call = function call(builder, args) {
    this.input = args[0];
    this.ids = (0, _normalizeIds2.default)(args[0], this.relation.relatedModelClass.getIdPropertyArray(), { arrayOutput: true });
    return true;
  };

  HasManyRelateOperation.prototype.queryExecutor = function queryExecutor(builder) {
    var _this2 = this;

    var patch = {};

    _lodash2.default.each(this.relation.relatedProp, function (relatedProp, idx) {
      patch[relatedProp] = _this2.owner[_this2.relation.ownerProp[idx]];
    });

    return this.relation.relatedModelClass.query().childQueryOf(builder).patch(patch).copyFrom(builder, /where/i).whereInComposite(this.relation.relatedModelClass.getFullIdColumn(), this.ids).modify(this.relation.filter);
  };

  HasManyRelateOperation.prototype.onAfterInternal = function onAfterInternal() {
    return this.input;
  };

  return HasManyRelateOperation;
}(_QueryBuilderOperation2.default);

exports.default = HasManyRelateOperation;
