// objection/lib/relations/hasMany/HasManyRelation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Relation2 = require('../Relation');

var _Relation3 = _interopRequireDefault(_Relation2);

var _HasManyInsertOperation = require('./HasManyInsertOperation');

var _HasManyInsertOperation2 = _interopRequireDefault(_HasManyInsertOperation);

var _HasManyRelateOperation = require('./HasManyRelateOperation');

var _HasManyRelateOperation2 = _interopRequireDefault(_HasManyRelateOperation);

var _HasManyUnrelateOperation = require('./HasManyUnrelateOperation');

var _HasManyUnrelateOperation2 = _interopRequireDefault(_HasManyUnrelateOperation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var HasManyRelation = function (_Relation) {
  (0, _inherits3.default)(HasManyRelation, _Relation);

  function HasManyRelation() {
    (0, _classCallCheck3.default)(this, HasManyRelation);
    return (0, _possibleConstructorReturn3.default)(this, _Relation.apply(this, arguments));
  }

  HasManyRelation.prototype.createRelationProp = function createRelationProp(owners, related) {
    var _this2 = this;

    var relatedByOwnerId = _lodash2.default.groupBy(related, function (related) {
      return related.$values(_this2.relatedProp);
    });

    _lodash2.default.each(owners, function (owner) {
      var ownerId = owner.$values(_this2.ownerProp);
      owner[_this2.name] = relatedByOwnerId[ownerId] || [];
    });
  };

  HasManyRelation.prototype.appendRelationProp = function appendRelationProp(owner, related) {
    owner[this.name] = this.mergeModels(owner[this.name], related);
  };

  HasManyRelation.prototype.insert = function insert(builder, owner) {
    return new _HasManyInsertOperation2.default(builder, 'insert', {
      relation: this,
      owner: owner
    });
  };

  HasManyRelation.prototype.relate = function relate(builder, owner) {
    return new _HasManyRelateOperation2.default(builder, 'relate', {
      relation: this,
      owner: owner
    });
  };

  HasManyRelation.prototype.unrelate = function unrelate(builder, owner) {
    return new _HasManyUnrelateOperation2.default(builder, 'unrelate', {
      relation: this,
      owner: owner
    });
  };

  return HasManyRelation;
}(_Relation3.default);

exports.default = HasManyRelation;
